### PR TITLE
Knn in oneapi interfaces

### PR DESCRIPTION
**Pull request by amgrigoriev**

### CODE REVIEW

File: cpp/oneapi/dal/algo/knn/detail/model_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/knn/common.hpp"
+
+namespace oneapi::dal::knn {
+
+class detail::model_impl: public base {
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/knn/detail/model_impl.hpp:21]:
```c++
// file: dal/detail/common.hpp
class pimpl_accessor {
    template <typename Object, typename ... Args>
    static auto make(Args&& ...args) {
        return Object{ std::forward<Args>(args)... };
    }
};


// file: common.hpp
namespace detail {
class model_impl;
} // namespace detail

class model {

};


// file: detail/model_impl.hpp
namespace detail {

class model_impl {
public:
    // If needed
    virtual foo() = 0;
};

} // namespace detail


// file: backend/interop_model.hpp
#include "detail/model_impl.hpp"
namespace backend {

class daal_interop_model : public model_impl {

};

} // namespace backend


// file: common.hpp
class model {
friend dal::detail::pimpl_accessor;
public:
    model();

private:
    explicit model(const std::shared_ptr<detail::model_impl>& impl);

    dal::detail::pimpl<detail::model_impl> impl_;
};


// file: common.cpp
#include "detail/model_impl.hpp"

class empty_model_impl : public detail::model_impl {};

// Default constructor
model::model() : impl_(new empty_model_impl{}) {}

model::model(const std::shared_ptr<detail::model_impl>& impl) : impl_(impl) {}


// file: backend/cpu/kernel.cpp
#include "backend/interop_model.hpp"

model train() {
    const auto model_impl = std::make_shared<daal_interop_model>(...);
    return dal::detail::pimpl_accessor::make<model>(model_impl);
}

```

File: cpp/oneapi/dal/algo/knn/backend/cpu/train_kernel_kd_tree.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    daal_knn::Parameter daal_parameter(desc.get_class_count(),
+        desc.get_neighbor_count(), desc.get_seed(), desc.get_data_use_in_model() ? daal_knn::doUse : daal_knn::doNotUse);
+
+    const daal_knn::ModelPtr model_ptr = 
+        daal_knn::ModelPtr(daal_knn::Model::create(column_count, NULL));
+    //if(model_ptr == nullptr)
+    //    bad alloc
+
+    model_ptr->impl()->setData<Float>(NumericTablePtr(daal_data.get()), desc.get_data_use_in_model());
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/knn/backend/cpu/train_kernel_kd_tree.cpp:67]:
```suggestion
    model_ptr->impl()->setData<Float>(daal_data, desc.get_data_use_in_model());
```

File: cpp/oneapi/dal/algo/knn/backend/cpu/train_kernel_kd_tree.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    model_ptr->impl()->setLabels<Float>(NumericTablePtr(daal_labels.get()), desc.get_data_use_in_model());
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/knn/backend/cpu/train_kernel_kd_tree.cpp:68]:
```suggestion
    model_ptr->impl()->setLabels<Float>(daal_labels, desc.get_data_use_in_model());
```

File: examples/oneapi/cpp/source/knn/knn_kd_tree_batch.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        1,
+        0,
+        1,
+        1
+    };
+
+    const auto data_table = dal::homogen_table{ row_count, column_count, data };
+    const auto label_table = dal::homogen_table{ row_count, 1, labels };
+
+    const auto knn_desc = dal::knn::descriptor<float, oneapi::dal::knn::method::kd_tree>();
```
Comment (rlnx) [inline examples/oneapi/cpp/source/knn/knn_kd_tree_batch.cpp:65]: I'd suggest to show as much parameters as possible

File: cpp/oneapi/dal/algo/knn/backend/cpu/train_kernel_kd_tree.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    // TODO: data is table, not a homogen_table. Think better about accessor - is it enough to have just a row_accessor?
+    const auto daal_data = interop::convert_to_daal_homogen_table(arr_data, row_count, column_count);
+    const auto daal_labels = interop::convert_to_daal_homogen_table(arr_labels, row_count, 1);
+
+
+    daal_knn::Parameter daal_parameter(desc.get_class_count(),
+        desc.get_neighbor_count(), desc.get_seed(), desc.get_data_use_in_model() ? daal_knn::doUse : daal_knn::doNotUse);
+
+    const daal_knn::ModelPtr model_ptr = 
+        daal_knn::ModelPtr(daal_knn::Model::create(column_count, NULL));
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/knn/backend/cpu/train_kernel_kd_tree.cpp:63]:
```suggestion
        daal_knn::Model::create(column_count, NULL);
```

File: cpp/oneapi/dal/algo/knn/backend/cpu/train_kernel_kd_tree.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+    // TODO: data is table, not a homogen_table. Think better about accessor - is it enough to have just a row_accessor?
+    const auto daal_data = interop::convert_to_daal_homogen_table(arr_data, row_count, column_count);
+    const auto daal_labels = interop::convert_to_daal_homogen_table(arr_labels, row_count, 1);
+
+
+    daal_knn::Parameter daal_parameter(desc.get_class_count(),
+        desc.get_neighbor_count(), desc.get_seed(), desc.get_data_use_in_model() ? daal_knn::doUse : daal_knn::doNotUse);
+
+    const daal_knn::ModelPtr model_ptr =
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/knn/backend/cpu/train_kernel_kd_tree.cpp:62]: I'd recommend to use `auto` if the type of right hand side expression is clear from context for the reader 
```suggestion
    const auto model_ptr = 
```

Ignore it, if you don't feel in the same way 😄

File: cpp/oneapi/dal/algo/knn/common.hpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    auto& set_data_use_in_model(bool value) {
+        set_data_use_in_model_impl(value);
+        return *this;
+    }
+};
+
+class model : public base {
+  friend dal::detail::pimpl_accessor;
+  public:
+    model(detail::model_impl *);
```
[the remark this answers is not in the document]
Comment (rlnx) [reply cpp/oneapi/dal/algo/knn/common.hpp:90]: Don't forget to mark constructor `explicit`

File: cpp/oneapi/dal/algo/knn/common.cpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+#include "oneapi/dal/algo/knn/common.hpp"
+#include "oneapi/dal/algo/knn/detail/model_impl.hpp"
+
+namespace oneapi::dal::knn {
+
+class detail::descriptor_impl : public base {
+  public:
+    std::int64_t class_count = 2;
+    std::int64_t neighbor_count = 1;
+    std::int64_t seed = 1;
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/knn/common.cpp:26]: Remove seed!

File: cpp/oneapi/dal/algo/knn/detail/model_impl.hpp (hunk 1 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/knn/detail/model_impl.hpp:21]: I like the idea of empty model. Will rework.

**Pull request closed without merging**